lib/backend/dir/impl.go

```diff
@@ -430,44 +419,207 @@ func (bk *Backend) ReleaseLock(token string) (err error) {
+	// Convert ["foo", "bar"] to "foo/bar"
+	raw := filepath.Join(bucket...)
+
+	// URL encode bucket from "foo/bar" to "foo%2Fbar".
+	flat := url.QueryEscape(raw)
+
+	return filepath.Join(bk.rootDir, flat)
 }
 
-// applyTTL assigns a given TTL to a file with sub-second granularity
-func (bk *Backend) applyTTL(dirPath string, key string, ttl time.Duration) error {
+// itemWithTTL returns a bucket item with TTL applied if the TTL was not
+// backend.Forever.
+func (bk *Backend) itemWithTTL(value []byte, ttl time.Duration) bucketItem {
+	bv := bucketItem{Value: value}
 	if ttl == backend.Forever {
-		return nil
+		return bv
 	}
-	expiryTime := bk.Clock().Now().Add(ttl)
-	bytes, _ := expiryTime.MarshalText()
-	return trace.ConvertSystemError(
-		ioutil.WriteFile(bk.ttlFile(dirPath, key), bytes, defaultFileMode))
+	bv.ExpiryTime = bk.Clock().Now().Add(ttl)
+	return bv
+
+}
+
+// isExpired checks if the bucket item is expired or not.
+func (bk *Backend) isExpired(bv bucketItem) bool {
+	if bv.ExpiryTime.IsZero() {
+		return false
+	}
+	return bk.Clock().Now().After(bv.ExpiryTime)
+}
+
+// bucket contains a set of keys that map to values and a TTL.
+type bucket struct {
+	// log is a structured component logger.
+	log *logrus.Entry
+
+	// file is the underlying file that the bucket represents.
+	file *os.File
+
+	// items is a set of key/value pairs that this bucket holds.
+	items map[string]bucketItem
+
+	// itemsUpdated is used to control if the items have been updated and should
+	// be written out to disk again.
+	itemsUpdated bool
 }
 
-// checkTTL checks if a given file has TTL and returns 'true' if it's expired
-func (bk *Backend) checkTTL(dirPath string, key string) (expired bool, err error) {
-	bytes, err := ioutil.ReadFile(bk.ttlFile(dirPath, key))
+// bucketItem is the "Value" part of a key/value pair.
+type bucketItem struct {
+	// Value is content of the key.
+	Value []byte `json:"value"`
+
+	// ExpiryTime is when this value will expire.
+	ExpiryTime time.Time `json:"expiry,omitempty"`
+}
+
+// openBucket will open a file, lock it, and then read in all the items in
+// the bucket.
+func openBucket(prefix string, openFlag int) (*bucket, error) {
+	// Open bucket with requested flags.
+	file, err := os.OpenFile(prefix, openFlag, defaultFileMode)
 	if err != nil {
-		if os.IsNotExist(err) { // no TTL
-			return false, nil
+		return nil, trace.ConvertSystemError(err)
+	}
+
+	// Lock the bucket so no one else can access it.
+	if err := utils.FSWriteLock(file); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// Read in all items from the bucket.
+	items, err := readIn(file)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return &bucket{
+		items: items,
+		file:  file,
+	}, nil
+}
+
+// Close will write out items (if requested), unlock file, and close it.
+func (b *bucket) Close() error {
+	var err error
+
+	// If the items were updated, write them out to disk.
+	if b.itemsUpdated {
+		err = writeOut(b.file, b.items)
```

Comment:
if you failed to write, still close the bucket, otherwise you will get a deadlock situation on busy file system